### ISSUE TITLE
Implement Adaptive Input Representations for Neural Language Modeling

### ISSUE THREAD

**Issue by jph00** [pytorch/pytorch#17313]
## 🚀 Feature

It would be great if you could implement [Adaptive Input Representations for Neural Language Modeling](https://arxiv.org/abs/1809.10853). This is, essentially, the same trick that PyTorch currently uses for adaptive softmax outputs, but applied to the input embeddings as well. In addition, it would be helpful to provide optional support for adaptive input and output weight tying.

## Motivation

PyTorch has already implemented adaptive representations for output. The paper shows that these representations are also useful for input. They allow faster training, and provide better accuracy. In addition, in some situations the paper shows it can be helpful to use weight-tying, which would allow this feature to be integrated nicely with the existing adaptive output functionality.

At the time the paper was written, adaptive input representations were the state of the art in language modeling.

## Alternatives

An alternative would be for this functionality to be moved in to higher level libs like fastai, but that would mean to get weight tying to work, either the existing pytorch adaptive output would have to be recreated in that library, or else that library would have to rely on using the internal data structures of the pytorch module - neither of these would be ideal.

**Comment by jph00** [comment]
Also, adaptive input was used in the Transformer XL paper, which is the current SoTA for LM.

**Comment by jph00** [comment]
Existing implementations:

- https://github.com/kimiyoung/transformer-xl/blob/2aa79e60ca653a025309c14c27672c055c9e65e0/pytorch/mem_transformer.py#L433
- https://github.com/AranKomat/adapinp/blob/master/AdapInp.py

**Comment by t-vi** [comment]
Indeed, probably next to the Adaptive Softmax would be a great place for it:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/adaptive.py
From what I can see @AranKomat 's implementation looks closer to PyTorch's Adaptive Softmax in style, but unless he himself puts in the PR, we'd need to independently implement it due to licensing.

**Comment by AranKomat** [comment]
@t-vi I deleted the license of the said implementation, so you can freely copy it to PyTorch. If that doesn't solve the issue, please let me know. On an unrelated note, it may be a good idea to prepare a simple PyTorch function to create a mapping based on the word frequency, including BOS, EOS and blank tokens, which is needed for adaptive i/o.

**Comment by t-vi** [comment]
@AranKomat Thank you! With you saying that you'd like it included to PyTorch that's cool.
I'm not quite sure where the best place for a mapping function would be - usually I'd expect something like that in torchtext and the like, but it seems that torchtext is a little less canonical than e.g. torchvision.
(Of course, I can't wait to see how @jph00 will use it, and then that would probably clear up, too.)

**Comment by jph00** [comment]
Embeddings are of course useful for *lots* of things other than text (text is just one popular use of high-cardinality categorical variables). I'm sure we'll add suitable preprocessing to fastai to make this more convenient to use, once its available, including text-specific mapping stuff in fastai.text.

**Comment by soumith** [comment]
@cpuhrsch this is relevant to your NLP discussions

**Comment by taufique74** [comment]
@t-vi Although it's been quite a while, I was wondering if this is still relevant and whether you're interested in a PR.
I implemented Adaptive Input inspired by @AranKomat 's implementation. There was an issue with @AranKomat's implementation for batch processing. I modified them so that this can replace `nn.Embedding` without any issue.
Here you can see my implementation (need some modification and cleanup), 
https://github.com/taufique74/AdaptiveIO/blob/0fd05695fd599b2706934959b76bb5ccb55521f3/model.py#L209

I trained several language models with `AdaptiveInput+AdaptiveSoftmax` and this is faster than `nn.Embedding+AdaptiveSoftmax` and consumes less memory if we tie the tail cluster weights.

**Comment by cpuhrsch** [comment]
cc @zhangguanheng66 who is the owner of torchtext

**Comment by zhangguanheng66** [comment]
@taufique74 Since this is more relevant to NLP applications, we might accept a PR for pytorch/text repo [here](https://github.com/pytorch/text).

**Comment by taufique74** [comment]
@zhangguanheng66 I thought of `torchtext` too, but couldn't find where this would fit. 
Also as @t-vi said, this should be with `Adaptive Softmax` since this is basically a hierarchical embedding layer, and will be replacing `torch.nn.Embedding` in its applications. So personally I think this should be in `torch/nn/modules/adaptive.py`.

If you still think this should be in the `torchtext`, please guide me where should I put this so that I can send a PR.

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 I thought of `torchtext` too, but couldn't find where this would fit.
> Also as @t-vi said, this should be with `Adaptive Softmax` since this is basically a hierarchical embedding layer, and will be replacing `torch.nn.Embedding` in its applications. So personally I think this should be in `torch/nn/modules/adaptive.py`.
> 
> If you still think this should be in the `torchtext`, please guide me where should I put this so that I can send a PR.

Make sense. I saw `AdaptiveLogSoftmaxWithLoss` in `adaptive.py` file. What's the difference with that?

**Comment by taufique74** [comment]
@zhangguanheng66 
The idea came from `AdaptiveSoftmax`, you can say `AdaptiveInput` is an extension of it. In language modeling, we often tie the weights of embedding and softmax layer. But we can't tie weights if we use `AdaptiveSoftmax`, since it breaks the layer into multiple clusters. So what if we convert the embedding layer into multiple clusters too? If we maintain the similar cluster size both in `AdaptiveInput` and `AdaptiveSoftmax`, we can tie the weights. This is one of the key idea behind `AdaptiveInput`.

Also, both of them exploit the fact that most of the corpus is covered by a small number of vocabulary (e.g., In Penn Treebank, 87% of the corpus is covered by 20% of the vocabulary). So if we allocate less space for tail clusters that contain the lookup table for less frequent words, we can save a lot of memory.

Even if we don't use `AdaptiveSoftmax` and tie the weights, we can save memory by allocating less space for tail clusters.
Consider a scenario where we have a vocabulary of size 1 million but only 100k covers 80% of the corpus. If we decide to use 300 as embedding dim, then only in embedding layer we have 300 million parameters. This is waste and we can reduce memory consumed in embedding layer by putting most frequent 100k words in head cluster. The rest 900k words go into multiple tail clusters and we allocate much smaller embedding dimensions for them.
Now, since we have saved up memory, we can fit more batches and eventually have less training time.

**Comment by taufique74** [comment]
@zhangguanheng66  @albanD 
Should I start working on a PR?

**Comment by zhangguanheng66** [comment]
@taufique74 Since the idea is similar to `AdaptiveLogSoftmaxWithLoss`, we can put it next to `torch.nn.modules.adaptive`. Ideally, we should put those future modules to `torchtext.modules`, if they are more relevant to the text domain, because torchtext is the extension of pytorch for NLP applications.

**Comment by taufique74** [comment]
@zhangguanheng66 As an early adopter of `torchtext`, I agree with you that the modules related to the text domain should be moved to `torchtext`.

>For this one, since the idea is similar to AdaptiveLogSoftmaxWithLoss, we can put it next to torch.nn.modules.adaptive. Do you have any ideas how non-text domain users can use this module?  

I'm not quite sure about the non-text domain use cases where we would need huge lookup tables. But since `AdaptiveInput` has trainable parameters and both the implementation and concept is pretty similar to `AdaptiveLogSoftmaxWithLoss`, I think putting them together makes much more sense in this case and might alleviate unnecessary confusion.
